### PR TITLE
Devops 177971256 memory analyzer window

### PR DESCRIPTION
**Pull request by yakloinsteak**

### CODE REVIEW

File: config/deploy/docker/lib/memory_analyzer.rb [hunk 1]
```diff
@@ -35,9 +39,9 @@ class MemoryAnalyzer
   # Number of metric values needed before we try to estimate RAM. If metrics
   # come in every 5 minutes, then this number represents
   # (MIN_SAMPLES * 5 / 60 / 24) days of data
-  MIN_SAMPLES = 2_000
+  MIN_SAMPLES = 288 # ~24 hours
 
-  MIN_RAM_MB = 800
+  MIN_RAM_MB = 600
```
Comment (yakloinsteak) [inline config/deploy/docker/lib/memory_analyzer.rb:44]: 800 seemed too high for our smallest containers

File: config/deploy/docker/lib/memory_analyzer.rb [hunk 2]
```diff
@@ -48,34 +52,52 @@ def run!
...
-      self.recommended_soft_limit_mb = (current_soft_limit_mb * ((_overall_stats.average + nsd * _overall_stats.stddev)/100.0)).ceil
+      # recommend some percentage above maximum utilization in recent past
+      self.recommended_hard_limit_mb = (current_soft_limit_mb * ((_overall_stats.maximum + 20.0) / 100.0)).ceil
+
+      self.recommended_soft_limit_mb =
+        begin
+          if task_definition_name.match?(/dj-(all|long)/)
+            # Percentage of maximum RAM
+            puts "[INFO][MEMORY_ANALYZER] Soft limit 95% of maximum"
+            ((current_soft_limit_mb * (_overall_stats.maximum / 100.0)) * 0.95).ceil
```
Comment (yakloinsteak) [inline config/deploy/docker/lib/memory_analyzer.rb:73]: Across all the deployments, I hope this eases the RAM contention, but moving to workoff containers is a much better fix I think.

File: config/deploy/docker/lib/roll_out.rb [hunk 3]
```diff
@@ -311,23 +327,22 @@ def _register_task!(name:, image:, cpu_shares: nil, soft_mem_limit_mb: 512, port
...
     ma.cluster_name         = self.cluster
     ma.task_definition_name = name
-    ma.scheduled_hard_limit_mb = hard_mem_limit_mb
-    ma.scheduled_soft_limit_mb = soft_mem_limit_mb
+    ma.bootstrapped_hard_limit_mb = hard_mem_limit_mb
+    ma.bootstrapped_soft_limit_mb = soft_mem_limit_mb
     ma.run!
 
-    # only doing this on staging for now to test the waters.
-    use_memory_analyzer = name.match?(/staging|vi/)
```
Comment (yakloinsteak) [inline config/deploy/docker/lib/roll_out.rb:319]: moved into the memory analyzer class

File: config/deploy/docker/lib/memory_analyzer.rb (hunk 1 above)
Comment (eanders) [reply config/deploy/docker/lib/memory_analyzer.rb:44]: I was finding that at that size I can ssh and open a console without hitting the hard limit.  I'd guess that's similar to booting up a DJ worker too, but I'm happy to try for a bit.  Especially if the soft limit is higher, maybe that was the issue.

File: config/deploy/docker/lib/memory_analyzer.rb [hunk 4]
```diff
@@ -48,34 +52,52 @@ def run!
       return
     end
 
-    unless _ram_settings_havent_changed_in_two_weeks?
-      puts "[INFO][MEMORY_ANALYZER] Cannot analyze RAM as it has changed in the past two weeks or there's not enough history"
-      self.recommended_hard_limit_mb = (current_values.current_hard_limit_mb || scheduled_hard_limit_mb).to_i
-      self.recommended_soft_limit_mb = (current_values.current_soft_limit_mb || scheduled_soft_limit_mb).to_i
+    unless _ram_settings_havent_changed_recently?
+      puts "[INFO][MEMORY_ANALYZER] Cannot analyze RAM as it has changed in the past day or there's not enough history"
+      self.recommended_hard_limit_mb = scheduled_hard_limit_mb
```
Comment (eanders) [inline config/deploy/docker/lib/memory_analyzer.rb:57]: is `scheduled_hard_limit_mb` now stored somewhere?  You renamed the attr_accessor above.

File: config/deploy/docker/lib/memory_analyzer.rb [hunk 5]
```diff
@@ -99,6 +121,14 @@ def recommended_hard_limit_mb
     @recommended_hard_limit_mb
   end
 
+  def scheduled_hard_limit_mb
+    (current_values.current_hard_limit_mb || bootstrapped_hard_limit_mb).to_i
+  end
```
Comment (eanders) [inline config/deploy/docker/lib/memory_analyzer.rb:126]: Oh here.

File: config/deploy/docker/lib/memory_analyzer.rb (hunk 1 above)
Comment (yakloinsteak) [reply config/deploy/docker/lib/memory_analyzer.rb:44]: I'll test opening a console

Comment (yakloinsteak) [reply config/deploy/docker/lib/memory_analyzer.rb:44]: got a console fine, but let me know if you have problems.

**Pull request merged**